Fix compound assignments for properties with different getter & setter types

jihndai opened the pull request:
<!--
Thank you for submitting a pull request!

Please verify that:
* [x] There is an associated issue in the `Backlog` milestone (**required**)
* [x] Code is up-to-date with the `main` branch
* [x] You've successfully run `gulp runtests` locally
* [x] There are new or updated unit tests validating the change

Refer to CONTRIBUTING.MD for more details.
  https://github.com/Microsoft/TypeScript/blob/main/CONTRIBUTING.md

** Please don't send typo fixes! **
Please don't send a PR solely for the purpose of fixing a typo, unless that
typo truly hurts understanding of the text. Each PR represents work for the
maintainers, and that work should provide commensurate value.

If you're interested in sending a PR, the issue tracker has many issues marked `help wanted`.
-->

Fixes #48292

Explanation on some changes:
* During binary expression checks, the function `checkPropertyAccessExpressionOrQualifiedName` would always check with the setter's type for the left operand when the operator was a compound assignment (++, +=, /=). This caused the typing error in the linked issue, because getters can be a subtype of setters. The fix was to check with the getter's type for compound assignments by default, and add a flag `writeOnly` to the `checkPropertyAccessExpressionOrQualifiedName` function to specify when to use the setter's type instead. This flag will be used when we want to perform type checking only on the assignment aspect of a compound assignment operator in a binary expression.
* ~~Why did a `skipReadWriteCheck` flag get added to the function `accessKind`? Prior to this fix, an expression `this.x` from a statement like `this.x += 1;` would return an `AccessKind.Write`, but we need it to return `AccessKind.ReadWrite` in order to know whether we need the getter's type instead of the setter's type. However, we also need to preserve the existing behavior for cases such as `function f(x: number) { x += 1; }` where we want to throw an "unused local" error for `x`. This existing behavior is enforced by tests such as `noUnusedLocals_writeOnly.ts`. So a flag was added to toggle between these two desired behaviors.~~

jakebailey commented:
> However, we also need to preserve the existing behavior for cases such as `function f(x: number) { x += 1; }` where we want to throw an "unused local" error for `x`

Is the current behavior actually desirable? I guess I would have expected that `x += 1` marks `x` as used. If you write

```ts
function f(x: number) {
    x;
}
```

`x` is marked as used, so it sounds to me like the new behavior is actually correct and you don't need to work around it.

jihndai commented:
I do think its a bit inconsistent that `x = x + 1` will mark x as read, while `x += 1` doesn't.

However, if we agree on changing the current behavior, should we open a separate issue/PR for this?

jakebailey commented:
I guess I'm confused; it seemed to me from your PR description that you went to some length to preserve that behavior, is there more to it than a simplification of this PR?

jihndai commented:
Nah, you understood it correctly: changing the current behavior of `accessKind` would result in not much more than a simplification of this PR.

My question was more concerned about whether or not a separate issue should be opened to change the current behavior of the function `accessKind`, given that it isn't the issue being addressed in #48292

jakebailey commented:
I'm not sure, maybe. I'm personally not against fixing inconsistencies like this outright. @RyanCavanaugh @DanielRosenwasser @andrewbranch may have stronger feelings about this one.

DanielRosenwasser commented:
It was very intentional that the left side of a compound assignment operators should not mark something as truly used for unused local checking. If it's not referenced elsewhere, then it's indicative of a logic bug. There's a hole in `x = x + 1` where we don't count usages that are solely used as part of initializing the same variable (have fun efficiently checking that with `y = x = x + 1`), but we shouldn't weaken the existing behavior.

I'm not familiar with the code. Is it "easy" to make `AccessKind.ReadWrite` not count as a read for unused-locals checking?

jihndai commented:
As in changing it so that variables will be considered as "used" only when it has an access kind of `AccessKind.Read` and nothing else? If that's the case, then yes, it seems simple to do.

However, after experimenting a bit with such a change, it seems to break existing behavior/test-cases such as:

### Case 1
```typescript
function foo(x: number) {
  console.log(x);
}

function bar() {
  let y = 0;
  foo(y++);
}
```
`y` will be marked as unused.

### Case 2 
```typescript
function f1 () {
    for(var i = 0, j= 0; ;i++) {}
}
```
Both `i` and `j` will be marked as unused, while the current expected behavior is that only `j` should be marked as unused.

jakebailey commented:
Personally, I think that ends up being even more inconsistent. My opinion is still that we should reevaluate our existing behavior, but given it wasn't as cut-and-dry as I expected, I guess it does deserve a new issue that links back to this PR with some examples, and I guess this PR will have to include the workarounds to preserve the current behavior in order until that issue is resolved.

jihndai commented:
I'm fine with proceeding with a workaround in the meantime.

To that end, the flag [in this line](https://github.com/microsoft/TypeScript/pull/50431/files#diff-1516c8349f7a625a2e4a2aa60f6bbe84e4b1a499128e8705d3087d893e01d367R5608) should help in preserving the current behavior with the unused local checks.

jihndai commented:
The behavior we want to preserve (and workaround) only applies when `accessKind` is used for unused locals check.

So some changes were made to separate that "unused locals check" rule into a separate function, which should be easier to understand than using a parameter flag.

sandersn commented:
@jakebailey I'm looking at old PRs and I wonder whether the changes since you last commented mean that the PR is ready to merge now. Can you take another look?

jakebailey commented:
It's probably ready to merge, but I still am of the opinion that it's a lot of code to maintain an inconsistency. But, that is not the fault of the PR; it seems like it has to be this way to preserve the existing behavior.

jakebailey commented on src/compiler/utilities.ts:
```diff
@@ -6703,21 +6713,38 @@ const enum AccessKind {
-    if (!parent) return AccessKind.Read;
 
-    switch (parent.kind) {
+    switch (parent?.kind) {
```
Why would parent ever be undefined here?

jakebailey replied:
Ah, didn't realize that the old code already had an early return.

jihndai replied:
Correct, as you saw this was inherited from the existing code that had a falsy check, even though the type suggests that it cannot be nullish.

Decided to run the tests without the optional chaining, and it seems that the tests from only this file `findAllRefsImportEqualsJsonFile.ts` are failing without it, so I guess we will keep this for now.

jakebailey replied:
> Correct, as you saw this was inherited from the existing code that had a falsy check, even though the type suggests that it cannot be nullish.

FWIW this is one case where our types lie; it's just too annoying to handle a missing parent because it only happens when the node is SourceFile.

jakebailey commented on src/compiler/utilities.ts:
```diff
@@ -6703,21 +6713,38 @@ const enum AccessKind {
+    const kind = accessKind(node);
+    const { parent } = node;
+
+    switch (parent?.kind) {
```
Similarly, here?

jakebailey commented:
@typescript-bot pack this

jakebailey commented:
Per #52829, we actually decided that we're going to go the direction of simplifying this so you don't need all of the extra code to distinguish accesses from usage.

I think this means this PR can be reduced down to just removing `writeOrReadWrite` and updating the doc for `accessKind`, but we're not going to make that change for 5.0.

jihndai commented:
Updated the changes to keep usage checks and access checks consistent. Had to update a lot of test cases that relied on marking variables/parameters as unused when using prefix/postfix unary operators.

Also deleted 4 codefix test files as they no longer seem to apply with the changes in this pr. These codefix test cases were introduced as part of #41624, which are: 
- `tests/cases/fourslash/codeFixUnusedIdentifier_postfix_operator1.ts`
- `tests/cases/fourslash/codeFixUnusedIdentifier_postfix_operator2.ts`
- `tests/cases/fourslash/codeFixUnusedIdentifier_prefix_operator1.ts`
 - `tests/cases/fourslash/codeFixUnusedIdentifier_prefix_operator2.ts`

jakebailey commented on src/compiler/checker.ts:
```diff
@@ -31165,7 +31167,7 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
     }
 
-    function checkPropertyAccessExpressionOrQualifiedName(node: PropertyAccessExpression | QualifiedName, left: Expression | QualifiedName, leftType: Type, right: Identifier | PrivateIdentifier, checkMode: CheckMode | undefined) {
+    function checkPropertyAccessExpressionOrQualifiedName(node: PropertyAccessExpression | QualifiedName, left: Expression | QualifiedName, leftType: Type, right: Identifier | PrivateIdentifier, checkMode: CheckMode | undefined, writeOnly?: boolean) {
```
Is the parameter here (and elsewhere) needed anymore now that we're changing `accessKind`?

jihndai replied:
Yes, it's needed in function `checkAssignmentOperatorWorker` to check assignability when the operator is a compound assignment. Without this flag, `checkPropertyAccessExpressionOrQualifiedName` will always use the getter's type when the operator is a compound assignment, as the setter's type is only really needed when checking assignability.

jakebailey commented:
Alright, skimming it, things look good. Let's run the suite, though.

@typescript-bot test this
@typescript-bot test top100
@typescript-bot user test this
@typescript-bot run dt
@typescript-bot perf test this

jakebailey commented:
Just to cross link them, this is effectively a reversal of the decisions of #51371 and #42422.

jakebailey reviewed (approved):
So, given the behavior change as-discussed, this PR looks good to me, though I think I want someone else to also take a look and double check.
